File: frontend/src/api/pipeline.ts
/**
 * API-Client für Pipeline-Automatisierung.
 * Endpoints: POST /generate/pipeline/run, GET /generate/pipeline/{id}, GET /generate/pipeline/{id}/stream
 */

import { API_BASE, apiClient } from "./client.js";

export interface PipelineRunRequest {
  prompt: string;
  image_provider_key?: string;
  image_params?: Record<string, unknown>;
  mesh_provider_key?: string;
  mesh_params?: Record<string, unknown>;
  enable_bgremoval?: boolean;
  bgremoval_provider_key?: string;
  enable_rigging?: boolean;
  rigging_provider_key?: string;
  enable_animation?: boolean;
  animation_provider_key?: string;
  motion_prompt?: string;
}

export type PipelineStepStatusValue = "pending" | "processing" | "done" | "failed" | "skipped";
export type PipelineRunStatusValue = "running" | "done" | "failed";

export interface PipelineStepStatus {
  step: string;
  job_id: string | null;
  status: PipelineStepStatusValue;
  result_url: string | null;
  error: string | null;
}

export interface PipelineRunStatus {
  pipeline_run_id: string;
  status: PipelineRunStatusValue;
  asset_id: string | null;
  steps: PipelineStepStatus[];
  created_at: string;
  updated_at: string | null;
  error: string | null;
}

export async function startPipelineRun(
  request: PipelineRunRequest
): Promise<PipelineRunStatus> {
  const { data } = await apiClient.post<PipelineRunStatus>(
    "/generate/pipeline/run",
    request
  );
  return data;
}

export async function getPipelineRunStatus(
  pipelineRunId: string
): Promise<PipelineRunStatus> {
  const { data } = await apiClient.get<PipelineRunStatus>(
    `/generate/pipeline/${pipelineRunId}`
  );
  return data;
}

/** Baut die SSE-Stream-URL für einen Pipeline-Run. */
export function getPipelineStreamUrl(pipelineRunId: string): string {
  const apiKey = import.meta.env.VITE_API_KEY as string | undefined;
  const url = `${API_BASE}/api/v1/generate/pipeline/${pipelineRunId}/stream`;
  return apiKey ? `${url}?api_key=${encodeURIComponent(apiKey)}` : url;
}

File: api/app/schemas/pipeline.py
"""Schemas für die Pipeline-Automatisierung (One-Click Full Pipeline)."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class PipelineRunRequest(BaseModel):
    """Anfrage für einen automatisierten Pipeline-Durchlauf."""

    prompt: str = Field(..., min_length=1, max_length=2000, description="Text-Prompt für die Bildgenerierung")
    image_provider_key: str = Field(default="picsart-default", description="Provider für Bildgenerierung")
    image_params: dict = Field(default_factory=dict, description="Parameter für Bildgenerierung")

    mesh_provider_key: str = Field(default="hunyuan3d-2", description="Provider für Mesh-Generierung")
    mesh_params: dict = Field(default_factory=dict, description="Parameter für Mesh-Generierung")

    enable_bgremoval: bool = Field(default=True, description="Background-Removal vor Mesh-Generierung")
    bgremoval_provider_key: str = Field(default="rembg-local", description="Provider für Background-Removal")

    enable_rigging: bool = Field(default=False, description="Rigging nach Mesh-Generierung")
    rigging_provider_key: str = Field(default="unirig", description="Provider für Rigging")

    enable_animation: bool = Field(default=False, description="Animation nach Rigging")
    animation_provider_key: str = Field(default="hy-motion", description="Provider für Animation")
    motion_prompt: str = Field(default="walk forward", description="Motion-Prompt für Animation")


class PipelineStepStatus(BaseModel):
    """Status eines einzelnen Pipeline-Schritts."""

    step: str
    job_id: UUID | None = None
    status: str  # pending | processing | done | failed | skipped
    result_url: str | None = None
    error: str | None = None


class PipelineRunStatus(BaseModel):
    """Status eines laufenden Pipeline-Durchlaufs."""

    pipeline_run_id: str
    status: str  # running | done | failed
    asset_id: UUID | None = None
    steps: list[PipelineStepStatus]
    created_at: datetime
    updated_at: datetime | None = None
    error: str | None = None


class PipelineRunResponse(BaseModel):
    """Antwort auf POST /generate/pipeline/run."""

    pipeline_run_id: str
    status: str
    asset_id: UUID | None = None
    steps: list[PipelineStepStatus]
    created_at: datetime
